compiler/src/dotty/tools/dotc/ast/MainProxies.scala

```diff
@@ -3,96 +3,244 @@ package ast
+    def mainMethods(scope: Tree, stats: List[Tree]): List[(Symbol, ParameterAnnotationss, DefaultValueSymbols, Option[Comment])] = stats.flatMap {
+      case stat: DefDef =>
+        val sym = stat.symbol
+        sym.annotations.filter(_.matches(defn.MainAnnot)) match {
+          case Nil =>
+            Nil
+          case _ :: Nil =>
+            val paramAnnotations = stat.paramss.flatMap(_.map(
+              valdef => valdef.symbol.annotations.filter(_.matches(defn.MainAnnotParameterAnnotation))
+            ))
+            (sym, paramAnnotations.toVector, defaultValueSymbols(scope, sym), stat.rawComment) :: Nil
+          case mainAnnot :: others =>
+            report.error(s"method cannot have multiple main annotations", mainAnnot.tree)
+            Nil
+        }
       case stat @ TypeDef(name, impl: Template) if stat.symbol.is(Module) =>
-        mainMethods(impl.body)
+        mainMethods(stat, impl.body)
       case _ =>
         Nil
     }
-    mainMethods(stats).flatMap(mainProxy)
+
+    // Assuming that the top-level object was already generated, all main methods will have a scope
+    mainMethods(EmptyTree, stats).flatMap(mainProxy)
   }
 
   import untpd._
-  def mainProxy(mainFun: Symbol)(using Context): List[TypeDef] = {
-    val mainAnnotSpan = mainFun.getAnnotation(defn.MainAnnot).get.tree.span
+  def mainProxy(mainFun: Symbol, paramAnnotations: ParameterAnnotationss, defaultValueSymbols: DefaultValueSymbols, docComment: Option[Comment])(using Context): List[TypeDef] = {
+    val mainAnnot = mainFun.getAnnotation(defn.MainAnnot).get
     def pos = mainFun.sourcePos
-    val argsRef = Ident(nme.args)
+    val mainArgsName: TermName = nme.args
+    val cmdName: TermName = Names.termName("cmd")
+
+    val documentation = new Documentation(docComment)
+
+    inline def lit(any: Any): Literal = Literal(Constant(any))
+
+    inline def some(value: Tree): Tree = Apply(ref(defn.SomeClass.companionModule.termRef), value)
+
+    def unitToValue(value: Tree): Tree =
+      val anonName = nme.ANON_FUN
+      val defdef = DefDef(anonName, List(Nil), TypeTree(), value)
+      Block(defdef, Closure(Nil, Ident(anonName), EmptyTree))
+
+    /**
+      * Creates a list of references and definitions of arguments, the first referencing the second.
+      * The goal is to create the
+      *   `val arg0: () => S = ...`
+      * part of the code. The first element of the tuple is a ref to `arg0`, the second is the whole definition.
+      */
+    def createArgs(mt: MethodType, cmdName: TermName): List[(Tree, ValDef)] =
+      mt.paramInfos.zip(mt.paramNames).zipWithIndex.map {
+        case ((formal, paramName), n) =>
+          val argName = mainArgsName ++ n.toString
+
+          val isRepeated = formal.isRepeatedParam
+
+          var argRef: Tree = Apply(Ident(argName), Nil)
+          var formalType = formal
+          if isRepeated then
+            argRef = repeated(argRef)
+            formalType = formalType.argTypes.head
```

Comment:
Maybe better to avoid the `var`s. I propose something like this
```scala
val (argRef, formalType, getterSym) = {
  val argRef0 = Apply(Ident(argName), Nil)
  if formal.isRepeatedParam then 
    (repeated(argRef0), formal.argTypes.head, defn.MainAnnotCommand_varargGetter)
  else (argRef0, formal, defn.MainAnnotCommand_argGetter)
}
```